### ISSUE TITLE
Discussion / Questions

### ISSUE THREAD

**Issue by emptyopen** [wsievern/BalletBot#2]

**Comment by emptyopen** [comment]
Regarding the memory vs disk topic: basically everything you store as a file is disk and "permanent", everything else is in memory and "temporary". Reading from disk is slow compared to reading from memory, but you generally have much more storage capacity on disk than in memory. Probably lots of good articles to google "disk vs memory". 

So when talking about the music files, those are always stored on the disk as files. The selected file is loaded into memory with the `load()` command [here](https://github.com/wsievern/BalletBot/blob/main/screens/playback/playback.py#L42). You could measure how long it takes to load it (measure time before and after, find difference) - with the current logic we are loading the selected file every time we hit play. You could load all the files into memory once upon starting the app, so long as there is sufficient memory to hold all the files. As the number of files increase however, this could become untenable. If you don't expect the number of songs to be very high (or more precisely the total size of the files to not be too big) it's probably best to just load it all once, for the fastest playback.

**Comment by wsievern** [comment]
This all makes sense. If I have a lot of music files, at a point would it make sense to store them on a cloud somewhere and have the app pull them in when needed?WSent from my iPhoneOn Dec 27, 2022, at 5:28 PM, Matt ***@***.***> wrote:﻿
Regarding the memory vs disk topic: basically everything you store as a file is disk and "permanent", everything else is in memory and "temporary". Reading from disk is slow compared to reading from memory, but you generally have much more storage capacity on disk than in memory. Probably lots of good articles to google "disk vs memory".
So when talking about the music files, those are always stored on the disk as files. The selected file is loaded into memory with the load() command here. You could measure how long it takes to load it (measure time before and after, find difference) - with the current logic we are loading the selected file every time we hit play. You could load all the files into memory once upon starting the app, so long as there is sufficient memory to hold all the files. As the number of files increase however, this could become untenable. If you don't expect the number of songs to be very high (or more precisely the total size of the files to not be too big) it's probably best to just load it all once, for the fastest playback.

**Comment by emptyopen** [comment]
This is exactly what a larger scale app would do. If the media is hardcoded into the app like it is currently, users would need to update the app in order to get new content. On the other hand, if the media is available through the cloud, commonly through a CNS system, you could simply update what's available in the cloud and that would be reflected in the app. This would be a pretty complex change so I don't think it's necessary for now.

**Comment by wsievern** [comment]
Yo, I just tried to commit some changes from my IDE... not sure if it worked. Did it go through for you?

**Comment by emptyopen** [comment]
Yeah I do see the updated code in `main`, unfortunately the new file isn't included in the project: `File '/Users/willsievern/Documents/GitHub/BalletBot/gillsans.ttf' not found`. You'll want to include any of these files in your data file, so maybe `data/fonts/gillsans.ttf`? So I can't really debug at the moment.

**Comment by emptyopen** [comment]
I would probably create new questions as new issue, or in this issue thread, as opposed to in the comments of the commit, since I don't think I get notified of those and will probably miss them. Or maybe you can @ me in the commit comments? If you want to upload uncommitted changes to Github for review, you can put your changes on a new branch with `git checkout -b <branch name>`, and `git push origin main`. Then you can create a pull request on Github, and tag me in specific line comments. In general it's definitely worth studying/practicing different git commands.

**Comment by wsievern** [comment]
For sure! Need to familiarize myself with the interface/commands. Thx for
your patience

W

**Comment by wsievern** [comment]
Just pushed a new branch... Not sure if I did it correctly. :/

I have been trying a few ways to update the bpm slider in the TapTempo
layout with the bpm from mem's loaded piece.  Still haven't cracked it, but
I've learned a lot about calling methods across classes in OOP.  From the KV
lang documentation
<https://kivy.org/doc/stable/api-kivy.lang.html#module-kivy.lang> it seemed
like it was possible to name an instance of TapTempo in BalletApp on
main.py, then reference said instance in a .kv file. As of the latest
branch, I've created the bpm slider in tap_tempo.kv, (id: bpm_slider) and
the method update_bpm_slider in tap_tempo.py, then mapping that function to
the back button in piece_selection.kv.  It feels close... but is not doing
anything. App is running though!

I also added all of the ballet exercises on the main page.  There are a
lot. Eventually, once I figure out the functionality of the Plie page (the
first exercise in a ballet class), I plan on having all the other pages
having the same interface.

W

**Comment by emptyopen** [comment]
Oh yeah I see the branch, so if you look for branches on the main page (left of the green button), then click the "new pull request" button, you should be able to open a pull request (basically "pulling" the branch into the main branch). Let me see if I can check out your branch in the mean time and see the same behavior.

**Comment by wsievern** [comment]
Yoooo I finally cracked the BPM slider value assignment issue. Will push
the branch soon.
